### ISSUE TITLE
Historic n-points over time query

### ISSUE THREAD

**Issue by leonerd** [prometheus/prometheus#9234]
(I admit that is a terrible title for this issue; I'm open to suggestions for a better one)

The prometheus `max_over_time()` query gives one number for the maximum of a vector over its entire history range. E.g.

The hottest temperature seen over the past 7 days was:

```
max_over_time(weather:temperature:degrees[7d])
```

You can optionally provide an "interval" in the notation, asking it to only calculate that value at the given interval:

```
max_over_time(weather:temperature:degrees[7d:1d])
```

This is veryalmost but not what I want. What I would like from the query is a timeseries of many values, showing the maximum value at the given time of the day, for each of the past 7 days. (Obviously the actual "1 day" and "7 day" parameters should be tuneable, they're just the easiest ones to explain). So, for this data, it would allow me to graph the maximum temperature for each moment of the day, across the past 7 days.

---

Background: I've been using Prometheus to store weather data, trying to generate useful graphs of measurements that follow a strong daily pattern, such as temperature or light level.

**Comment by leonerd** [comment]
By way of perhaps better explanation here is a screenshot of what I can get out currently:
![temperature](https://user-images.githubusercontent.com/350941/130359994-19647435-5016-4425-b719-91471c7f9ad5.png)

It has evaluated the value just once per 24h (1d) period of time, and drawn a flat yellow line between them all. Instead, I would like to see something like the following badly-drawn mockup:
![temperature-7daymax](https://user-images.githubusercontent.com/350941/130360094-a95c98b9-7c8f-47c4-ae11-416487e1ddcf.png)

With this kind of query, I'd then be able to draw a graph of the current temperature, displayed within a context of the min/max over the past 7 days.

**Comment by SuperQ** [comment]
I think this should be possible with subquery syntax.

**Comment by leonerd** [comment]
An initial attempt at subquerying doesn't appear to work. E.g. I've added

```yaml
      - record: 'weather:temperature:max7d'
        expr: 'max_over_time(weather:temperature:avg5m[7d:1d])'
      - record: 'weather:temperature:min7d'
        expr: 'min_over_time(weather:temperature:avg5m[7d:1d])'
```

and the numbers I get out don't vary at all across a single 24h period. I believe this comes from this part of https://prometheus.io/blog/2019/01/28/subquery-support/:

> rate(http_requests_total[5m]) is executed from start=<now>-30m to end=<now>, at a resolution of 1m. Note that start time is aligned independently with step of 1m (aligned steps are 0m 1m 2m 3m ...).

where my "start" times are being quantized to a per-day basis.

**Comment by LeviHarrison** [comment]
#9114 adds support for aligning subqueries to the evaluation time, ignoring the performance optimization. Maybe that helps?

**Comment by beorn7** [comment]
Indeed, I think this is another use case that #9114 addresses.

**Comment by roidelapluie** [comment]
I agree with @beorn7 , this will be addressed by #9114

**Comment by krajorama** [comment]
Hello from the bug scrub: closing as subset of #9114.
